nemo_text_processing/g2p/models/ctc_g2p.py

```diff
@@ -0,0 +1,417 @@
+# Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import string
+from dataclasses import dataclass
+from typing import Any, Dict, List, Optional, Union
+
+import torch
+from hydra.utils import instantiate
+from nemo_text_processing.g2p.data.ctc_g2p import CTCG2PBPEDataset
+from omegaconf import DictConfig, ListConfig, OmegaConf, open_dict
+from pytorch_lightning import Trainer
+from torch import nn
+from transformers import AutoConfig, AutoModel, AutoTokenizer
+
+from nemo.core.classes.common import PretrainedModelInfo
+from nemo.core.classes.modelPT import ModelPT
+from nemo.utils import logging
+
+try:
+    from nemo.collections.asr.losses.ctc import CTCLoss
+    from nemo.collections.asr.metrics.wer_bpe import WERBPE, CTCBPEDecoding, CTCBPEDecodingConfig
+    from nemo.collections.asr.metrics.wer import word_error_rate
+    from nemo.collections.asr.models import EncDecCTCModel
+    from nemo.collections.asr.parts.mixins import ASRBPEMixin
+
+    ASR_AVAILABLE = True
+except (ModuleNotFoundError, ImportError) as e:
+    ASR_AVAILABLE = False
+
+
+__all__ = ['CTCG2PModel']
+
+
+@dataclass
+class CTCG2PConfig:
+    train_ds: Optional[Dict[Any, Any]] = None
+    validation_ds: Optional[Dict[Any, Any]] = None
+
+
+class CTCG2PModel(ModelPT, ASRBPEMixin):
+    """
+    CTC-based grapheme-to-phoneme model.
+    """
+
+    # @property
+    # def input_types(self) -> Optional[Dict[str, NeuralType]]:
+    #     return {
+    #         "input_ids": NeuralType(('B', 'T'), TokenIndex()),
+    #         "attention_mask": NeuralType(('B', 'T'), MaskType(), optional=True),
+    #         "labels": NeuralType(('B', 'T'), LabelsType()),
+    #     }
+    #
+    # @property
+    # def output_types(self) -> Optional[Dict[str, NeuralType]]:
+    #     return {"loss": NeuralType((), LossType())}
+
+    def __init__(self, cfg: DictConfig, trainer: Trainer = None):
+        self.world_size = 1
+        if trainer is not None:
+            self.world_size = trainer.num_nodes * trainer.num_gpus
```

Comment:
Doesn't the PTL trainer have a `world_size`? Do we have best practices for what to use?

Reply:
This is used across NMT and ASR as far as I can tell
https://github.com/NVIDIA/NeMo/blob/phoneme_asr_ev/nemo/collections/asr/models/ctc_models.py#L47

Reply:
Gotcha. Maybe we should sync sometime with the team as a whole and figure out a standard 😅